### ISSUE TITLE
EventEmitter

### ISSUE THREAD

**Issue by Jonas74** [epha/dynamo#6]
How would event emitter compare to promises regarding flow of queries?  May it be an option, to create an event emitter per table or tables, which controls the table, knows the schema, encryption and handles all references (like populate)?

**Comment by meaku** [comment]
## Streams 
Using streams:

```javascript 
var itemsStream = dynamo.QueryStream({ catId: "cats" }),
      populateStream = new PopulateStream("user"), //populates all user attributes
      decryptStream = new DecryptStream();  

itemStream.pipe(decryptStream).pipe(populateStream).pipe(res); 
``` 

## Hooks / Events

Another option using events similar to how mongoose solves it: http://mongoosejs.com/docs/api.html#schema_Schema-post

Hooks and additional schema changes could also be bundled as a plugin: http://mongoosejs.com/docs/plugins.html

__Per Model__ 

```javascript 

var User = new Schema({
   name: String,
   age: Number,
   key: String
});

User.pre('save', function () {
    this.key = encrypt(this.key);
});

User.post('fetch', function () {
    this.key = decrypt(this.key);
});
```

__Global__ 

```javascript 
dynamo.post("fetch", function(Model, item) {
   return decrypt(item);
});
```

That's quite similar to my suggestion in issue #3. 

I think we should split the library here. _dynamo_ takes care of low level dynamodb operations and adds convenience like streams and promises. Transformation, validation & population should happen a level above using a module that has a schema + model and knows how to use dynamodb. 

Still thinking... ;)

**Comment by meaku** [comment]
@Jonas74 is the event approach i described similar to the robot used in https://github.com/epha/download/blob/master/app.js

**Comment by meaku** [comment]
I've just found [dynagoose](https://github.com/automategreen/dynamoose) - mongoose for DynamoDB - which is sort of what i was proposing.

**Comment by meaku** [comment]
There is another well maintained library: https://github.com/totemstech/node-dynamodb
And there is a streams-lib for dynamo: https://www.npmjs.com/package/dynamo-streams

**Comment by meaku** [comment]
# Conclusion

## Model instance methods vs. Batch methods
Dynagoose has a nice API for operations on single model instances, but batch methods don't really fit in there. Unfortunately the more advanced features like `population`, `hooks` and _EmbeddedDocuments_ are not implemented yet. 

## Schema 
I would use an existing library to convert from/to DynamoDB-JSON format by type checking, so a schema is not needed anymore for this case. But a schema would still be a good addition for things like _encryption/decryption_, _transformation_ and _population_.

## Promises 
Promises are great and they should be the main interface, when calling the methods. 
That's the way [dynamite](https://github.com/Medium/dynamite) works as well.

## Streams 
Streams are a great addition and we could find some inspiration at  https://www.npmjs.com/package/dynamo-streams

## Events/Hooks
I think we need some kind of event/hook system to perform things like decryption/population. The definition might be part of our schema. 

We should start working on a project testing different libs and styles to find out which approach works best for us.

**Comment by Jonas74** [comment]
Thx for the nice post. I agree on working with different libs to find the best fit. What I like is the syntax sugar you proposed. All libraries are missing additional function, that would ease the actual work with the database (e.g. epha/data/bin/dynamo_import). 
```javascript
bot.use( localDB ).upload( file ).into( table )

bot.on('progress', function(data) {});
bot.on('done', function(data) {});
bot.on('error', function(err) {});
```
The generic statement, that promises are great, does not necessarily imply, that we should use them for aws. E.g. the dynamite lib doesn't significantly improve the handling, besides some syntax sugar. I rather would do something similar to the above.
```javascript
bot.use( localDB ).get( hash, range ).from( table )

bot.on('done', function(raw,dataItem) {});
bot.on('error', function(raw,err) {});
```

**Comment by meaku** [comment]
We should definitely add syntax sugar. I really like the syntax of the bot example. 

Regarding promises and streams/events: Promises improve the code dramatically imho when it comes to nesting of function calls. If data is streamable we should stick with streams/events. Both types are useful for different use cases and therefor complementary rather then competing.

**Comment by Jonas74** [comment]
I agree. Maybe [cypher]( http://neo4j.com/developer/cypher-query-language/) gives some more inspiration regarding syntax.
